Add new runners to integrate Robolectric with junit-platform, junit jupiter

utzcoz opened the pull request:
1. `:runner:common` is the base and common module for other runner integrations.
2. `:runner:junit-platform` is the runner to integrate Robolectric with junit-platform.
3. `:runner:junit-jupiter` is the runner to integrate Robolectric with junit-jupiter by extension mode like https://github.com/apter-tech/junit5-robolectric-extension.

This version is the initial version written by GitHub Copilot with multiple models like Gemini 3 Pro(Preview), Claude Sonnet 4.5, and GPT-5.1-Codex with my vibe-coding practice. And I will update it manually based on feedback, thoughts, and more tests.

For https://github.com/robolectric/robolectric/issues/3477.
For https://github.com/robolectric/robolectric/issues/3397.
For https://github.com/robolectric/robolectric/issues/10727.

utzcoz commented on runner/kotest/src/main/kotlin/org/robolectric/kotest/RobolectricSpec.kt:
```diff
@@ -0,0 +1,60 @@
+abstract class RobolectricSpec(body: FunSpec.() -> Unit = {}) : FunSpec(body) {
+  init {
+    extensions(RobolectricExtension())
+  }
```
Done.

utzcoz commented:
/gemini review

utzcoz commented:
/gemini review

utzcoz commented:
Hi @hoisie @brettchabot , I am doing the prototype of JUnit platform and JUnit Jupiter supporting, and also for general testing framework integration https://github.com/robolectric/robolectric/issues/10797. It's first step made by GitHub Copilot's vibe coding approach. And now I want to receive your feedback and suggestions for this prototype, and I will polish it manually based on review and suggestions.
If it can work, we can continue to extract existing Robolectric test runner to a new junit4 runner module, and :robolectric module only leaves shadow and critical parts of framework.

nishatoma commented on runner/junit-jupiter/src/test/kotlin/org/robolectric/shadows/ShadowActivityTest.kt:
```diff
@@ -0,0 +1,43 @@
+import org.robolectric.junit.jupiter.RobolectricExtension
+import org.robolectric.testapp.TestActivity
+
+@ExtendWith(RobolectricExtension::class)
```
Semi-relevant to this annotation: I know this is probably our of scope for this PR, is there any plan to also support something similar to `AndroidJUnit4` runner? Or for example `@Parametrized...Extension`?

utzcoz replied:
I think not. From JUnit5/Jupiter, it doesn't use runner explicit like AndroidJUnit4.

utzcoz commented:
/gemini review

MGaetan89 commented on runner/common/build.gradle.kts:
```diff
@@ -0,0 +1,74 @@
+java {
+  sourceSets["main"].java.setSrcDirs(listOf("src/main/kotlin"))
+  sourceSets["test"].java.setSrcDirs(listOf("src/test/kotlin"))
+}
```
I think that this is not necessary. If I remove it, I still see the files as part of the project.

MGaetan89 commented on runner/common/build.gradle.kts:
```diff
@@ -0,0 +1,74 @@
+  sourceSets["test"].java.setSrcDirs(listOf("src/test/kotlin"))
+}
+
+kotlin { compilerOptions { jvmTarget = JvmTarget.JVM_21 } }
```
Why not set this in the `RoboKotlinModulePlugin`?

MGaetan89 commented on runner/junit-jupiter/build.gradle.kts:
```diff
@@ -0,0 +1,70 @@
+  implementation(libs.junit.platform.launcher)
+  implementation(libs.junit.platform.engine)
+  implementation(libs.junit4)
+  compileOnly(libs.findbugs.jsr305)
```
Same question here too

MGaetan89 commented on runner/junit-platform/build.gradle.kts:
```diff
@@ -0,0 +1,68 @@
+configurations {
+  testCompileClasspath {
+    attributes.attribute(
+      Attribute.of("com.android.build.api.attributes.BuildTypeAttr", String::class.java),
```
And same suggestion here

MGaetan89 commented on settings.gradle.kts:
```diff
@@ -53,6 +53,9 @@ include(
   ":robolectric",
+  ":runner:common",
+  ":runner:junit-platform",
+  ":runner:junit-jupiter",
```
```suggestion
  ":runner:junit-jupiter",
  ":runner:junit-platform",
```

MGaetan89 reviewed (commented):
Here's a first set of comments.
I've only looked at the Gradle files for the moment.

MGaetan89 commented on build-logic/convention/src/main/java/org/robolectric/gradle/AttributeNames.kt:
```diff
@@ -0,0 +1,9 @@
+class AttributeNames {
+  companion object {
+    // AGP's attribute BuildTypeAttr
+    const val BUILD_TYPE_ATTR = "com.android.build.api.attributes.BuildTypeAttr"
```
```suggestion
object AttributeNames {
  // AGP's attribute BuildTypeAttr
  const val BUILD_TYPE_ATTR = "com.android.build.api.attributes.BuildTypeAttr"
```

utzcoz replied:
Oh, good catch.

MGaetan89 reviewed (commented):
Looks good for the Gradle-specific changes. I'll start looking at each module soon

utzcoz commented on runner/junit-jupiter/src/test/kotlin/org/robolectric/shadows/ShadowActivityTest.kt:
```diff
@@ -0,0 +1,43 @@
+    val startedIntent = shadowOf(activity).nextStartedActivity
+    assertThat(startedIntent).isEqualTo(intent)
+  }
+
```
@nishatoma Add more tests related to configuration changes. You can check latest commit.

utzcoz commented on runner/common/src/main/kotlin/org/robolectric/runner/common/DefaultRobolectricIntegration.kt:
```diff
@@ -0,0 +1,215 @@
+          sdkContexts[tempContext.sdk.apiLevel]
+        } else {
+          null
+        }
```
Done.

rnett commented:
It would be great if this could be extended slightly (it's already 99% of the way there) to support running generic non-test snippets (i.e. lambdas).  This would make it useful for REPLs, compose preview renderers, etc, that run outside of the the context of tests.

utzcoz commented:
@hoisie @brettchabot What do you think about this PR? I need feedbacks.

utzcoz commented:
@brettchabot I added a runner-design.md and you can review it.

utzcoz commented:
> It would be great if this could be extended slightly (it's already 99% of the way there) to support running generic non-test snippets (i.e. lambdas). This would make it useful for REPLs, compose preview renderers, etc, that run outside of the the context of tests.

@rnett Could you provide some examples that you want?

rnett commented:
@utzcoz ideally something like just `Roboelectric.setup()` or a version that takes a lambda. My end use case is a MCP tool that can render compose previews - I already have a REPL, and it works fine for Compose Desktop where there is a rendering backed, but it doesn't on Android for obvious reasons, and I was hoping to be able to use Roboelectric to solve that (without dealing with test harnesses, etc).

utzcoz committed:
feat(runner): add shared runner modules with JUnit Platform/Jupiter engines and standalone runtime

Add reusable runner infrastructure for integrating Robolectric with test
frameworks beyond the classic JUnit 4 runner:

- :runner:common: sandbox and class-lifecycle management, discovery
  helpers, parameter resolution, SDK selection, and the canonical
  execution policy shared by both Jupiter entry points -- classes without
  @BeforeAll/@AfterAll run in isolated per-method environments honoring
  method-level @Config; classes with class lifecycle share one
  environment, with @RobolectricSdkTest as the explicit per-SDK isolation
  opt-in and fail-fast on implicit divergence.
- :runner:junit-platform: a custom JUnit Platform TestEngine with sandbox
  bootstrap, per-test lifecycle, and multi-SDK discovery.
- :runner:junit-jupiter: RobolectricExtension for the standard Jupiter
  engine (Jupiter's instance is a placeholder; tests re-invoke on a
  sandbox-loaded twin) and a dedicated Jupiter TestEngine whose discovery
  skips @ExtendWith(RobolectricExtension) classes, honors @Disabled,
  fails loudly on unsupported method kinds, and expands multi-SDK
  variants.
- RobolectricRuntime.launch { sdk = 34 }: a framework-free Android
  environment for non-test consumers like REPLs and preview renderers.

Tests are split into per-target source sets (test, jupiterExtensionTest,
engineFixtures) so each test runs only under its owning task. The
architecture is documented in runner-design.md and the integration
matrix.

Signed-off-by: utzcoz <utzcoz@outlook.com>

MGaetan89 commented on gradle/libs.versions.toml:
```diff
@@ -113,6 +113,7 @@ androidx-biometric = "1.1.0"
 androidx-constraintlayout = "2.2.1"
 androidx-core = "1.19.0"
 androidx-fragment = "1.8.9"
+androidx-lifecycle = "2.8.7"
```
```suggestion
androidx-lifecycle = "2.11.0"
```

utzcoz replied:
Done.

MGaetan89 commented on runner/common/build.gradle.kts:
```diff
@@ -0,0 +1,65 @@
+fun androidStubsJar(): ConfigurableFileCollection {
+  val androidStubsVersion = libs.versions.androidstubs.get()
+  val androidHome = System.getenv("ANDROID_HOME")
+  require(!androidHome.isNullOrBlank()) { "ANDROID_HOME environment variable not set or blank." }
```
It is recommended to throw `GradleException` in Gradle build script. Here you use an `IllegalArgumentException`

utzcoz replied:
Done.

MGaetan89 commented on runner/junit-jupiter/build.gradle.kts:
```diff
@@ -0,0 +1,121 @@
+//   jupiterExtensionTest classpath but in no task's testClassesDirs, so they execute only when a
+//   driving test launches them.
+val engineFixtures: SourceSet =
+  sourceSets.create("engineFixtures") {
```
Can't you use `register` instead of `create`?

utzcoz replied:
Done.

MGaetan89 commented on runner/junit-jupiter/build.gradle.kts:
```diff
@@ -0,0 +1,121 @@
+  }
+
+val jupiterExtensionTestSourceSet: SourceSet =
+  sourceSets.create("jupiterExtensionTest") {
```
Same here

utzcoz replied:
Done.

MGaetan89 commented on runner/junit-jupiter/build.gradle.kts:
```diff
@@ -0,0 +1,121 @@
+  named("engineFixturesCompileOnly") { extendsFrom(getByName("testCompileOnly")) }
+  named("jupiterExtensionTestImplementation") { extendsFrom(getByName("testImplementation")) }
+  named("jupiterExtensionTestCompileOnly") { extendsFrom(getByName("testCompileOnly")) }
+  named("jupiterExtensionTestRuntimeOnly") { extendsFrom(getByName("testRuntimeOnly")) }
```
```suggestion
  named("engineFixturesImplementation") { extendsFrom(named("testImplementation")) }
  named("engineFixturesCompileOnly") { extendsFrom(named("testCompileOnly")) }
  named("jupiterExtensionTestImplementation") { extendsFrom(named("testImplementation")) }
  named("jupiterExtensionTestCompileOnly") { extendsFrom(named("testCompileOnly")) }
  named("jupiterExtensionTestRuntimeOnly") { extendsFrom(named("testRuntimeOnly")) }
```

utzcoz replied:
Done.

MGaetan89 commented on runner/junit-jupiter/build.gradle.kts:
```diff
@@ -0,0 +1,121 @@
+fun androidStubsJar(): ConfigurableFileCollection {
+  val androidStubsVersion = libs.versions.androidstubs.get()
+  val androidHome = System.getenv("ANDROID_HOME")
+  require(!androidHome.isNullOrBlank()) { "ANDROID_HOME environment variable not set or blank." }
```
Same remark here about using `GradleException`

utzcoz replied:
Done.

MGaetan89 commented on runner/junit-platform/build.gradle.kts:
```diff
@@ -0,0 +1,55 @@
+fun androidStubsJar(): ConfigurableFileCollection {
+  val androidStubsVersion = libs.versions.androidstubs.get()
+  val androidHome = System.getenv("ANDROID_HOME")
+  require(!androidHome.isNullOrBlank()) { "ANDROID_HOME environment variable not set or blank." }
```
Here too

utzcoz replied:
Done.

MGaetan89 reviewed (commented):
Some comments/questions about the Gradle setup

utzcoz commented:
I want to bump the priority of this PR's reviewing, and I hope it can land at 4.18 release if possible. I think many people want it although it is experimental.
